src/ophyd_async/epics/pmac/_utils.py

```diff
@@ -0,0 +1,98 @@
+from dataclasses import dataclass
+
+import numpy as np
+import numpy.typing as npt
+from scanspec.core import Slice
+
+from ophyd_async.epics.motor import Motor
+
+# PMAC durations are in milliseconds
+# We must convert from scanspec durations (seconds) to milliseconds
+# PMAC motion program multiples durations by 0.001
+# (see https://github.com/DiamondLightSource/pmac/blob/afe81f8bb9179c3a20eff351f30bc6cfd1539ad9/pmacApp/pmc/trajectory_scan_code_ppmac.pmc#L241)
+# Therefore, we must divide scanspec durations by 10e-6
+TICK_S = 0.000001
+
+
+@dataclass
+class Trajectory:
+    positions: dict[Motor, np.ndarray]
+    velocities: dict[Motor, np.ndarray]
+    user_programs: npt.NDArray[np.int32]
+    durations: npt.NDArray[np.float64]
+
+    @classmethod
+    def from_slice(
+        cls,
+        slice: Slice,
+    ) -> "Trajectory":
+        """Parse a trajectory with no gaps from a slice.
+
+        :param slice: Information about a series of scan frames along a number of axes
+        :param ramp_up_duration: Time required to ramp up to speed
+        :param ramp_down: Booleon representing if we ramp down or not
+        :returns Trajectory: Data class representing our parsed trajectory
+        :raises RuntimeError: Slice must have no gaps and a duration array
+        """
+        if slice.duration is None:
+            raise RuntimeError("Slice must have a duration")
```

Comment:
You can do
https://github.com/bluesky/ophyd-async/blob/a67ef2aca0a8170acd7491cd0139c7125507c418/src/ophyd_async/core/_utils.py#L391-L404
```suggestion
        duration = error_if_none(slice.duration, "Slice must have a duration")
```